csrc/optimization/optimization_pass.h

```diff
@@ -0,0 +1,85 @@
+// clang-format off
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2023-present NVIDIA CORPORATION & AFFILIATES.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+// clang-format on
+#pragma once
+
+#include <ir/interface_nodes.h>
+
+#include <mutex>
+
+namespace nvfuser::optimization {
+
+using FusionPass = std::function<void(Fusion*)>;
+
+//! [experimental API]
+//! Base class to unify optimization pass APIs.
+//! OptimizationPass is functional and defines the granularity of mutation
+//! passes that is used to compose OptimizationGroups
+class TORCH_CUDA_CU_API OptimizationPass {
+ public:
+  virtual void run(Fusion*) = 0;
+  virtual std::string name() = 0;
+  virtual ~OptimizationPass() = default;
+};
+
+//! [experimental API]
+//! Base class to unify optimization group APIs.
+//! OptimizationGroup composes optimization passes that is used at certain stage
+//! in the runtime system. OptimizationGroup can be turned on/off
```

Comment:
Should we make the on/off flag a per-pass flag, instead of per-pass group? For example, if I want to debug X, and in order to debug X, I would like to disable consecutive cast simplification. But I do want other passes to run, so that I don't get noisy fusions.

Reply:
I think this is again a design decision.

I'm not a big fan of having a flag per-pass. Imagine that we might have complicated optimization group, where a certain pass is dependent on preceding passes.
I think having the granularity at the group level is a `safer` bet?!

Reply:
Today, we only have one pass, so either way, they are the same. However, in the future, I am likely to add a new pass that replaces same size as same `Val*`. For example, if we have `T1[I1{i1}] + T2[I2{i2}]`, then we should replace it as `T1[I1{i1}] + T2[I2{i1}]`. This is a totally independent thing w.r.t. this cast optimization, and I don't see any reason why we have to disable this pass if we just don't want cast optimization.

In general, if there are multiple passes depending on each other, then I think we should enable/disable them as a whole. But for passes not depending on each other, I don't see why we are having a all or nothing choice here.

I think probably we should add another level of hierarchy, that is, keep this `OptimizationGroup` as is, but `PreSegmenter` shouldn't be an optimization group, but a collective of multiple optimization groups where each group can be enabled/disabled as a whole?